csharp/src/Microsoft.ML.OnnxRuntime/DisposableNamedOnnxValue.cs

```diff
@@ -125,15 +125,108 @@ internal static DisposableNamedOnnxValue CreateFromOnnxValue(string name, IntPtr
+                        sequence.Add(CreateFromOnnxValue(string.Empty, nativeOnnxValueSeq, allocator));
+                        NativeMethods.OrtReleaseValue(nativeOnnxValueSeq);
+                    }
+                    return new DisposableNamedOnnxValue(name, sequence, null);
+
+                case OnnxType.ONNX_TYPE_MAP:
+                    IntPtr typeAndShape = IntPtr.Zero;
+                    IntPtr nativeOnnxValueMapKeys = IntPtr.Zero;
+                    IntPtr nativeOnnxValueMapValues = IntPtr.Zero;
+                    TensorElementType elemType = TensorElementType.DataTypeMax;
+                    NativeApiStatus.VerifySuccess(NativeMethods.OrtGetValue(nativeOnnxValue, 0, allocator, out nativeOnnxValueMapKeys));
+                    NativeApiStatus.VerifySuccess(NativeMethods.OrtGetValue(nativeOnnxValue, 1, allocator, out nativeOnnxValueMapValues));
+                    NativeApiStatus.VerifySuccess(NativeMethods.OrtGetTensorShapeAndType(nativeOnnxValueMapKeys, out typeAndShape));
+
+                    elemType = NativeMethods.OrtGetTensorElementType(typeAndShape);
+                    if (typeAndShape != IntPtr.Zero)
+                    {
+                        NativeMethods.OrtReleaseTensorTypeAndShapeInfo(typeAndShape);
+                    }
+                    switch (elemType)
+                    {
+                        case TensorElementType.Int64:
+                            return DisposableNamedMapFromNativeMap<Int64, float>(string.Empty, nativeOnnxValueMapKeys, nativeOnnxValueMapValues);
+                        case TensorElementType.String:
+                            return DisposableNamedMapFromNativeMap<string, float>(string.Empty, nativeOnnxValueMapKeys, nativeOnnxValueMapValues);
+                        default:
+                            throw new NotSupportedException("Map of element type: " + elemType + " is not supported");
+                    }
+                default:
+                    throw new NotSupportedException("OnnxValueType : " + onnxValueType + " is not supported");
+            }
+        }
 
-        private static DisposableNamedOnnxValue NameOnnxValueFromNativeTensor<T>(string name, IntPtr nativeOnnxValue)
+        private static DisposableNamedOnnxValue DisposableNamedOnnxValueFromNativeTensor<T>(string name, IntPtr nativeOnnxValue)
         {
             NativeOnnxTensorMemory<T> nativeTensorWrapper = new NativeOnnxTensorMemory<T>(nativeOnnxValue);
             DenseTensor<T> dt = new DenseTensor<T>(nativeTensorWrapper.Memory, nativeTensorWrapper.Dimensions);
             return new DisposableNamedOnnxValue(name, dt, nativeTensorWrapper);
         }
 
+        private static DisposableNamedOnnxValue DisposableNamedMapFromNativeMap<K,V>(string name, IntPtr nativeOnnxValueKeys, IntPtr nativeOnnxValueValues)
+        {
+            var map = new Dictionary<K, V>();
```

Comment:
you need to create a disposable dictionary similar to disposable list I guess... otherwise, how will the user dispose the underlying native memory?

Reply:
When a model output signature is something like **seq<map<K, V>>**, the InferenceSession.Run() call will return **DisposableList< DisposableList< DisposableNamedOnnxValue< object >>>**, where the object is of type **System.Collections.Generic.Dictionary<K,V>**.  

It looks like **DisposableList** will dispose all the **DisposableNamedOnnxValue** objects it contains, which will release the native data buffers. I believe the managed types will be automatically disposed when the **DisposableNamedOnnxValue** is deallocated, and therefore we shouldn't need to  explicitly dispose the map structures (which are created in managed memory using data copied from native memory). Let me know if you meant something different above.